chain/src/main/java/org/ethereum/beacon/chain/GhostHeadFunction.java

```diff
@@ -0,0 +1,222 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * The beacon chain fork choice rule is a hybrid that combines justification and finality with
+ * Latest Message Driven (LMD) Greediest Heaviest Observed SubTree (GHOST). For more info check <a
+ * href="https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#beacon-chain-fork-choice-rule">Beacon
+ * chain fork choice rule</a>
+ */
+public class GhostHeadFunction implements HeadFunction {
+
+  private final BeaconBlockStorage blockStorage;
+  private final BeaconTupleStorage tupleStorage;
+  private final SpecHelpers specHelpers;
+  private final int CHILDREN_SEARCH_LIMIT = Integer.MAX_VALUE;
+  private final Map<Bytes48, Attestation> attestationCache = new HashMap<>();
+  private final Map<UInt64, Set<Bytes48>> validatorSlotCache = new HashMap<>();
+  private BeaconState state;
+
+  public GhostHeadFunction(
+      BeaconChain beaconChain, BeaconChainStorage chainStorage, SpecHelpers specHelpers) {
+    this.tupleStorage = chainStorage.getBeaconTupleStorage();
+    this.blockStorage = chainStorage.getBeaconBlockStorage();
+    this.specHelpers = specHelpers;
+    this.state = tupleStorage.getCanonicalHead().getState();
+    Flux.from(beaconChain.getBlockStatesStream())
+        .doOnNext(beaconTuple -> state = beaconTuple.getState())
+        .subscribe();
+  }
+
+  @Override
+  public Optional<BeaconBlock> update() {
+    Hash32 curHead = blockStorage.getCanonicalHead();
+    UInt64 lastJustifiedSlot = state.getJustifiedSlot();
+    Optional<BeaconTuple> justifiedTuple =
+        blockStorage
+            .getSlotCanonicalBlock(lastJustifiedSlot.getValue())
+            .map(tupleStorage::get)
+            .orElseThrow(() -> new RuntimeException("No justified head found"));
+    BeaconBlock newHead =
+        justifiedTuple
+            .map(this::lmd_ghost)
+            .orElseThrow(() -> new RuntimeException("No justified head found"));
+    if (!newHead.getHash().equals(curHead)) {
+      blockStorage.reorgTo(newHead.getHash());
+      return Optional.of(newHead);
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  // def lmd_ghost(store, start):
+  //    validators = start.state.validator_registry
+  //    active_validators = [validators[i] for i in
+  //                         get_active_validator_indices(validators, start.state.slot)]
+  //    attestation_targets = [get_latest_attestation_target(store, validator)
+  //                           for validator in active_validators]
+  //    def get_vote_count(block):
+  //        return len([target for target in attestation_targets if
+  //                    get_ancestor(store, target, block.slot) == block])
+  //
+  //    head = start
+  //    while 1:
+  //        children = get_children(head)
+  //        if len(children) == 0:
+  //            return head
+  //        head = max(children, key=get_vote_count)
+  private BeaconBlock lmd_ghost(BeaconTuple startTuple) {
```

Comment:
Maybe move this one and other functions defined below to `SpecHelpers`? As their source is the spec. `store` parameter could be replaced with `Function<Hash32, List<BeaconBLock>>`.